bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/WlanThermoHandler.java

```diff
@@ -0,0 +1,190 @@
+                    checkConnectionAndUpdate();
+                } else {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                            "WlanThermo not found under given address.");
+                }
+            } catch (URISyntaxException | ExecutionException | TimeoutException e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        "Could not connect to WlanThermo at " + config.getIpAddress() + ": " + e.getMessage());
+            } catch (InterruptedException e) {
+                logger.debug("Connection check interrupted. {}", e.getMessage());
+            }
+        } else {
+            pull();
+        }
+    }
+
+    protected boolean doPost(String endpoint, String json) throws InterruptedException {
+        try {
+            URI uri = config.getUri(endpoint);
+            int status = httpClient.POST(uri).content(new StringContentProvider(json), "application/json")
+                    .timeout(5, TimeUnit.SECONDS).send().getStatus();
+            if (status == 401) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        "No or wrong login credentials provided. Please configure username/password for write access to WlanThermo!");
+                return false;
+            } else if (status != 200) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        "Failed to update channel on device, Statuscode " + status + " on URI " + uri.toString());
+                logger.debug("Payload sent: {}", json);
+                // Still continue to try next channel
+                return true;
+            } else {
+                updateStatus(ThingStatus.ONLINE);
+                return true;
+            }
+        } catch (TimeoutException | ExecutionException | URISyntaxException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "Failed to update channel on device: " + e.getMessage());
+            return false;
+        }
+    }
+
+    protected <T> T doGet(String endpoint, Class<T> object) throws InterruptedException, WlanThermoInputException {
+        try {
+            String json = httpClient.GET(config.getUri(endpoint)).getContentAsString();
+            logger.debug("Received at {}: {}", endpoint, json);
+            return requireNonNull(gson.fromJson(json, object));
+        } catch (URISyntaxException | ExecutionException | TimeoutException | WlanThermoException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "Update failed: " + e.getMessage());
+            for (Channel channel : thing.getChannels()) {
+                updateState(channel.getUID(), UnDefType.UNDEF);
+            }
+            throw new WlanThermoInputException();
```

Comment:
Always pass the cause exceptions into new exceptions you throw.
Please add such a constructor is one doesn't already exist.
```suggestion
            throw new WlanThermoInputException(e);
```

Also, I don't think that a WlanThermoInputException is the correct thing to throw here if you get a TimeoutException. The WlanThermoInputException message implies incorrect data but a Timeout is not related to that at all, so the exception message would be misleading. Consider having the `doGet` method throw WlanThermoException instead so you can throw broader types of exceptions.

Reply:
A Timeout would however result in invalid input data for the binding.
Changed it anyways.